src/Component.php

```diff
@@ -2,15 +2,111 @@
+use Keboola\StorageApi\Metadata;
+use Keboola\StorageApi\TableExporter;
 
 class Component extends BaseComponent
 {
     public function run(): void
     {
-        // @TODO implement
+        try {
+            /** @var Config $config */
+            $config = $this->getConfig();
+            $client = new Client(['token' => $config->getToken(), 'url' => $config->getUrl()]);
+            $tokenInfo = $client->verifyToken();
+            if (count($tokenInfo['bucketPermissions']) > 1) {
+                throw new UserException('The token must have read-only permissions to a single bucket only.');
+            }
+            $bucket = array_keys($tokenInfo['bucketPermissions'])[0];
+            if ($tokenInfo['bucketPermissions'][$bucket] !== 'read') {
+                throw new UserException(
+                    sprintf('The token must have read-only permissions to the bucket "%s".', $bucket)
+                );
+            }
+            if ($config->getAction() === 'run') {
+                $this->extract($client, $config, $bucket);
+            } elseif ($config->getAction() === 'list') {
+                echo \GuzzleHttp\json_encode([
+                    'tables' => $this->listTables($client, $bucket),
+                ]);
+            } else {
+                throw new UserException(sprintf('Unknown action "%s"', $config->getAction()));
+            }
+        } catch (ClientException $e) {
+            throw new UserException($e->getMessage());
+        }
+    }
+
+    private function extract(Client $client, Config $config, string $bucket) : void
+    {
+        if (empty($config->getTableName())) {
+            throw new UserException("The tableName parameter must be provided.");
+        }
+        $this->getLogger()->info(sprintf('Processing table "%s".', $config->getTableName()));
+        $exporter = new TableExporter($client);
+        $tableId = $bucket . '.' . $config->getTableName();
+        if (!empty($config->getChangedSince())) {
+            $options['changedSince'] = $config->getChangedSince();
+        } else {
+            $options = [];
+        }
+        $exporter->exportTable(
+            $tableId,
+            $this->getDataDir() . '/out/tables/' . $config->getTableName() . '.csv',
+            $options
+        );
+        $tableInfo = $client->getTable($tableId);
+        $metadata = new Metadata($client);
+        $columnMetadata = [];
+        foreach ($tableInfo['columns'] as $column) {
+            $colMetadata = $this->filterMetadata(
+                $metadata->listColumnMetadata($tableId . '.' . $column)
+            );
+            if ($colMetadata) {
+                $columnMetadata[$column] = $colMetadata;
+            }
+        }
+        $tableMetadata = $this->filterMetadata($metadata->listTableMetadata($tableId));
+        $options = new OutTableManifestOptions();
+        if ($columnMetadata) {
+            $options->setColumnMetadata($columnMetadata);
+        }
+        if ($tableMetadata) {
```

Comment:
`tableMetadata` i `columnMetadata` bych udělal volitelný a defaultně na false. 
`columnMetadata` jsou asi jednoduchý, ale u `tableMetadata` je dost zbytečný přenášet, jaká konfigurace/token vytvořila tabulku, když to bude jinej projekt a potenciálně i jinej region (a `tableMetadata` přepíše částečně i docker runner). do UI bych ani jednu z těch možností nedával.

co myslíš?

Reply:
nebo bych začal úplně bez metadat.

Reply:
dal jsem je optional (mazat se mi to uplne nechce :), v UI to byt nemusi
to, jaka konfigurace a komponenta tabulku vytvorila, se stejne prepise, takze se tim neni potreba zalamovat, ale muze tam byt neco jinyho